src/ucp/am/eager.c

```diff
@@ -250,104 +250,108 @@ ucp_proto_eager_zcopy_single_init(const ucp_proto_init_params_t *init_params)
     return ucp_proto_single_init(&params);
 }
 
-static void ucp_proto_eager_zcopy_send_func_add_usr_hdr(ucp_request_t *req,
-                                                        ucp_lane_index_t lane,
-                                                        uct_iov_t *iov,
-                                                        size_t *iovcnt)
+static void ucp_proto_eager_am_zcopy_add_footer(ucp_request_t *req,
+                                                ucp_lane_index_t lane,
+                                                uct_iov_t *iov, size_t *iovcnt)
 {
-    ucp_mem_desc_t *user_hdr_desc = req->send.msg_proto.am.reg_desc;
-    size_t user_hdr_size          = req->send.msg_proto.am.header_length;
-    ucp_md_index_t md_idx         = ucp_ep_md_index(req->send.ep, lane);
+    size_t user_hdr_size = req->send.msg_proto.am.header_length;
+    ucp_mem_desc_t *user_hdr_desc;
+    ucp_md_index_t md_idx;
+
+    if (user_hdr_size == 0) {
+        return;
+    }
+
+    user_hdr_desc = req->send.msg_proto.am.reg_desc;
+    md_idx        = ucp_ep_md_index(req->send.ep, lane);
 
     ucp_add_uct_iov_elem(iov, user_hdr_desc + 1, user_hdr_size,
                          user_hdr_desc->memh->uct[md_idx], iovcnt);
 }
 
 static ucs_status_t
-ucp_proto_eager_zcopy_send_func(ucp_request_t *req,
-                                const ucp_proto_single_priv_t *spriv,
-                                const uct_iov_t *iov)
+ucp_proto_eager_am_zcopy_single_send_func(ucp_request_t *req,
+                                          const ucp_proto_single_priv_t *spriv,
+                                          uct_iov_t *iov)
 {
-    size_t iovcnt = 0;
-    uct_iov_t zcopy_iov[2];
+    size_t iovcnt = 1;
     ucp_am_hdr_t hdr;
 
     ucp_am_fill_header(&hdr, req);
 
-    ucp_add_uct_iov_elem(zcopy_iov, iov->buffer, iov->length, iov->memh,
-                         &iovcnt);
-
-    if (req->send.msg_proto.am.header_length != 0) {
-        ucp_proto_eager_zcopy_send_func_add_usr_hdr(req, spriv->super.lane,
-                                                    zcopy_iov, &iovcnt);
-    }
+    ucp_proto_eager_am_zcopy_add_footer(req, spriv->super.lane, iov, &iovcnt);
```

Comment:
can we ensure the underlying transport supports 2 IOVs, and if it doesn't we need to disqualify the protocol during initialization